test/jdk/javax/swing/JProgressBar/TestProgressBarBorder.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8224261
+ * @key headful
+ * @library ../regtesthelpers
+ * @build Util
+ * @summary Verifies JProgressBar border is not painted when border
+ *          painting is set to false
+ * @run main TestProgressBarBorder
+ */
+
+import java.io.File;
+import java.awt.Graphics;
+import java.awt.image.BufferedImage;
+import javax.imageio.ImageIO;
+import javax.swing.JComponent;
+import javax.swing.JProgressBar;
+import javax.swing.SwingUtilities;
+import javax.swing.UIManager;
+import javax.swing.UnsupportedLookAndFeelException;
+
+import static java.awt.image.BufferedImage.TYPE_INT_RGB;
+
+public class TestProgressBarBorder {
+    private static JProgressBar progressBar;
+    private static volatile boolean isImgEqual;
+    private static BufferedImage borderPaintedImg;
+    private static BufferedImage borderNotPaintedImg;
```

Comment:
All these could be local variables.

Reply:
`boolean` variable is changed to local variable. Others are used in EDT and other method, so kept it as class variables.

Reply:
> `boolean` variable is changed to local variable. Others are used in EDT and other method, so kept it as class variables.

Do they need to be? I mean you can do everything on EDT, even throw the exception from there. It could something like this:

```java
public class TestProgressBarBorder {
    public static void main(String[] args) throws Exception {
        for (UIManager.LookAndFeelInfo laf :
                UIManager.getInstalledLookAndFeels()) {
            if (!laf.getName().contains("Nimbus") && !laf.getName().contains("GTK")) {
                continue;
            }
            System.out.println("Testing LAF: " + laf.getName());
            SwingUtilities.invokeAndWait(() -> test(laf));
        }
    }

    private static void test(UIManager.LookAndFeelInfo laf) {
        setLookAndFeel(laf);

        JProgressBar progressBar = createProgressBar();
        progressBar.setBorderPainted(true);
        BufferedImage withBorder = paintToImage(progressBar);
        progressBar.setBorderPainted(false);
        BufferedImage withoutBorder = paintToImage(progressBar);

        boolean equal = Util.compareBufferedImages(withBorder, withoutBorder);
        if (equal) {
            try {
                ImageIO.write(withBorder, "png", new File("withBorder.png"));
                ImageIO.write(withoutBorder, "png", new File("withoutBorder.png"));
            } catch (IOException ignored) {}
            throw new RuntimeException("JProgressBar border is painted when border\n" +
                                       " painting is set to false");
        }
    }

    private static JProgressBar createProgressBar() {
        JProgressBar progressBar = new JProgressBar();
        progressBar.setSize(100, 50);
        progressBar.setValue(0);
        progressBar.setStringPainted(true);
        return progressBar;
    }
```

Because the exception is thrown from EDT, the exception thrown from main is `InvocationTargetException`, yet the CI will still show you the cause. If you want to avoid it, you can return the boolean value from the `test` method.

Reply:
> Do they need to be? I mean you can do everything on EDT, even throw the exception from there. 

Yeah, everything can be done on EDT but what is the advantage of that?
I mean as per current test, swing UI components are accessed on EDT and then the comparison is done on main thread. 

Curious to know what we achieve extra by doing the way you suggested above?

Reply:
> > Do they need to be? I mean you can do everything on EDT, even throw the exception from there.
> 
> Yeah, everything can be done on EDT but what is the advantage of that?

Less code, all the variables are local in one method, no need to care about any synchronisation whatsoever.

> I mean as per current test, swing UI components are accessed on EDT and then the comparison is done on main thread.
> 
> Curious to know what we achieve extra by doing the way you suggested above?

What do we achieve by separating painting and comparing?

There's no way that's the only right one. Both do the same. Yet a single test-method which does everything is easier to understand: set look-and-feel, get a progress bar, paint it with and without border and, finally, compare the images. I'm always for this style.

When dealing with multiple threads you have to pass objects between threads; to do so, you store them as static fields — it somewhat complicates the data flow.

However, there's still one advantage to this — a clearer exception about the failure.

Reply:
Sounds good.
Updated.